src/conn/conn_api.c

```diff
@@ -1969,23 +1970,50 @@ __wt_diagnostic_asserts_config(WT_SESSION_IMPL *session, const char *cfg[])
     conn = S2C(session);
 
     WT_RET(__wt_config_gets(session, cfg, "diagnostic_asserts", &cval));
-#ifdef HAVE_DIAGNOSTIC
+
     if (cval.len > 0)
         WT_RET_MSG(session, EINVAL,
-          "WiredTiger has been compiled with HAVE_DIAGNOSTIC=1 and diagnostic_asserts are always "
+          "WiredTiger has been compiled with HAVE_DIAGNOSTIC=1 and all assertions are always "
           "enabled. This cannot be configured.");
-    FLD_SET(conn->debug_flags, WT_CONN_DEBUG_DIAGNOSTIC_ASSERTS);
+    conn->runtime_asserts_flags = WT_DIAG_ALL;
+
+    return (0);
+}
 #else
-    if (WT_STRING_MATCH("off", cval.str, cval.len))
-        FLD_CLR(conn->debug_flags, WT_CONN_DEBUG_DIAGNOSTIC_ASSERTS);
-    else if (WT_STRING_MATCH("on", cval.str, cval.len))
-        FLD_SET(conn->debug_flags, WT_CONN_DEBUG_DIAGNOSTIC_ASSERTS);
-    else if (cval.len > 0)
-        WT_RET_MSG(session, EINVAL, "Invalid value provided for diagnostic_asserts");
-#endif
+/*
+ * __wt_diagnostic_asserts_config --
+ *     Set diagnostic assertions configuration.
+ */
+int
+__wt_diagnostic_asserts_config(WT_SESSION_IMPL *session, const char *cfg[])
+{
+    static const WT_NAME_FLAG assert_types[] = {{"all", WT_DIAG_ALL},
+      {"concurrent_access", WT_DIAG_CONCURRENT_ACCESS},
+      {"data_validation", WT_DIAG_DATA_VALIDATION}, {"invalid_op", WT_DIAG_INVALID_OP},
+      {"out_of_order", WT_DIAG_OUT_OF_ORDER}, {"panic", WT_DIAG_PANIC},
+      {"slow_operation", WT_DIAG_SLOW_OPERATION}, {"visibility", WT_DIAG_VISIBILITY}};
 
+    WT_CONNECTION_IMPL *conn;
+    WT_CONFIG_ITEM cval, sval;
+    WT_DECL_RET;
+    const WT_NAME_FLAG *ft;
+    uint32_t flags;
+
+    conn = S2C(session);
+
+    WT_RET(__wt_config_gets(session, cfg, "diagnostic_asserts", &cval));
+
+    flags = 0;
```

Comment:
As per our offline discussion, I think this would work when passing an empty list as the diagnostic_asserts config even when not explicitly listed as a choice (see api_data.py:530) but best to be confirmed with unit tests.

Reply:
Very good catch! When I was testing this I caught I case I'd missed the first time round.
I've updated this function to parse config items properly. The results are as follows:
```
Compiled with HAVE_DIAG = 0
    (Not provided)                          // no assertions are turned on
    diagnostic_asserts=[]                   // no assertions are turned on
    diagnostic_asserts=[all, out_of_bounds] // all and out_of_bounds are turned on

Compiled with HAVE_DIAG = 1
    (Not provided)                          // All assertions are on
    diagnostic_asserts=[]                   // All assertions are on
    diagnostic_asserts=[all, out_of_bounds] // WT_ERR returned as we can't configure asserts in diagnostic mode

```

Reply:
Instead of looping through the full list of assertions here, could we instead check for an empty list as a valid configuration and return the error message otherwise? Similar to the earlier iteration of this config parsing function.